lib/pymedphys/_experimental/streamlit/utilities/iview/ui.py

```diff
@@ -0,0 +1,341 @@
+        seconds=offset_to_apply
+    )
+    icom_datasets["time"] = icom_datasets["datetime"].dt.round("ms").dt.time
+
+    try:
+        icom_datasets = _angles.make_icom_angles_continuous(icom_datasets)
+    finally:
+        if advanced_mode:
+            beam_on_mask = _utilities.expand_border_events(
+                np.diff(icom_datasets["meterset"]) > 0
+            )
+            beam_shade_min = -200 * beam_on_mask
+            beam_shade_max = 200 * beam_on_mask
+
+            icom_datasets["beam_shade_min"] = beam_shade_min
+            icom_datasets["beam_shade_max"] = beam_shade_max
+
+            beam_on_chart = (
+                alt.Chart(icom_datasets)
+                .mark_area(
+                    fillOpacity=0.1, strokeOpacity=0.3, stroke="black", fill="black"
+                )
+                .encode(x="datetime:T", y="beam_shade_min:Q", y2="beam_shade_max:Q")
+            )
+
+            device_angle_chart = (
+                beam_on_chart
+                + (
+                    alt.Chart(icom_datasets)
+                    .transform_fold(
+                        ["gantry", "collimator", "turn_table"], as_=["device", "angle"]
+                    )
+                    .mark_line(point=True)
+                    .encode(
+                        x="datetime:T",
+                        y=alt.Y("angle:Q", axis=alt.Axis(title="Angle (degrees)")),
+                        color="device:N",
+                        tooltip=["time:N", "device:N", "angle:Q"],
+                    )
+                    .properties(title="iCom Angle Parameters")
+                    .interactive(bind_y=False)
+                )
+            ).configure_point(size=10)
+
+            st.altair_chart(device_angle_chart, use_container_width=True)
+
+    if advanced_mode:
+        field_size_chart = (
+            alt.Chart(icom_datasets)
+            .transform_fold(["length", "width"], as_=["side", "size"])
+            .mark_line()
+            .encode(
+                x="datetime:T",
+                y="size:Q",
+                color="side:N",
+                tooltip=["time:N", "side:N", "size:Q"],
+            )
+            .properties(title="iCom Field Size")
+            .interactive(bind_x=False)
+        )
+        st.altair_chart(field_size_chart, use_container_width=True)
+
+        st.write(icom_datasets)
+
+    midnight = (
+        icom_datasets["datetime"]
+        .iloc[0]
+        .replace(hour=0, minute=0, second=0, microsecond=0)
+    )
+
+    icom_datasets["seconds_since_midnight"] = (
+        icom_datasets["datetime"] - midnight
+    ).dt.total_seconds()
+    database_table["seconds_since_midnight"] = (
+        database_table["datetime"] - midnight
+    ).dt.total_seconds()
+
+    icom_datasets["x_lower"] = icom_datasets["centre_x"] - icom_datasets["width"] / 2
+    icom_datasets["x_upper"] = icom_datasets["centre_x"] + icom_datasets["width"] / 2
```

Comment:
DRY
You are accessing the same value twice.
It's not going to save much code or improve performance but it might read slightly better to have an x_half_width variable.  Same for y_half_height

Reply:
Yup certainly. I specifically remember letting this one go at the time.

My hope with the workflow with this project was to do the following:

* Write experimental code designed to see if it meets clinical goals
  * Have this be allowed to be "experimental", in every sense of the word.
  * Have this go through a few iterations with staff here to see if it meets the latent need (<- this is the step I am still up to).
  * Potentially this code is "throw out code", so make sure not to sink too much development or review time into it.
  * Of course, if review time is spent, that review time should not be wasted, so make sure to implement the feedback if relevant.
  * This stage is optimising for minimal time being sunk on code that will be deleted, with the goal to quickly narrow down on exactly what is needed.
* Once latent need being met is confirmed, then write end to end tests so that refactoring and code improvement can be undergone without needing to redo the manual latent needs tests above.
* Once e2e tests are in place, step through moving sub-sections out of experimental, going through the full code review process knowing that any changes requested can be undergone with a degree of confidence that the code being changed has an appropriate test harness.

---

In saying that though, this comment here has been addressed in the following commit:

[`799f084` (#1416)](https://github.com/pymedphys/pymedphys/pull/1416/commits/799f0846821892c327d6e049983f17a84a0605f7)

Reply:
It's probably worth jumping on a video call with @Matthew-Jennings yourself and myself so that we can nut out how best to handle all this sort of stuff. Is there a good time @sjswerdloff and @Matthew-Jennings where you guys might be keen to jump on a video call, say hi to each other (I don't think you two have met), and also sort this out?

...maybe beer should be included?